### PR TITLE
Datasets

### PR DESCRIPTION
**Pull request by ms2892**
Hey Reviewers

Just added the classes to easily get the datasets for all 3 tasks - classification, regression and reinforcement learning.

Added the documentation as well and for easy use, please refer to the dataset_readme.md file. It contains all the steps required to get the dataset objects or download the datasets. 

Please make sure to test it by creating the jupyter notebook in the home directory where all the folders (features,models,utils etc) are. 

I have tested it personally and would like you guys to give it a try as well.

Sincerely
Mohd Sadiq

### CODE REVIEW

File: utils/datasets.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
+import torch
+import torchvision
+from torchvision import datasets, models, transforms
```
Comment (AugustasMacijauskas) [inline utils/datasets.py:3]: datasets, models imports not used

File: utils/datasets.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+from torch.utils.data import Dataset,DataLoader
```
Comment (augustasio) [inline utils/datasets.py]: DataLoader import not used

File: utils/datasets.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+import torch.nn as nn
```
Comment (augustasio) [inline utils/datasets.py]: Import not used

File: utils/datasets.py (hunk 1 above)
Comment (augustasio) [inline utils/datasets.py]: datasets, models imports not used

File: utils/datasets.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+import numpy as np
+import math
```
Comment (augustasio) [inline utils/datasets.py]: Import not used

File: utils/datasets.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+import os
+import matplotlib.pyplot as plt
```
Comment (augustasio) [inline utils/datasets.py]: Import not used

File: utils/datasets.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+        else:
+            
+            # Found the folder.
+            logging.info('Folder for UCI Mushroom found in the data folder. If the files are corrupted please delete the folder at location ../data/UCI_Mushroom and re-run this command')
+ 
+ 
+# This was meant for Debugging        
+if __name__=='__main__':
+    dataset = Datasets()
+    mnist = dataset.download_UCI()
```
Comment (augustasio) [inline utils/datasets.py]: Misleading variable name "mnist"

File: utils/datasets.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+                Returns the output values of the generating function with Gaussian Noise applied
+            Args:
+                x       (np.array): Array of Input Variables
+                sigma   (float)   : Variance of Gaussian Noise to be applied
+            Output:
+                (np.array)  : Numpy Array containing output values for each input values
+        """
+        
+        # Calculate the outputs with Gaussian Noise applied
+        y = x + 0.3 * np.sin(2*np.pi*(x+np.random.normal(0,sigma,x.shape))) + 0.3 * np.sin(4*np.pi*(x + np.random.normal(0,sigma,x.shape))) + np.random.normal(0,sigma,x.shape)
```
Comment (augustasio) [inline utils/datasets.py]: I am not sure about this, but maybe we should consider using the same epsilon everywhere, as the equation in the paper seems like it's the same thing everywhere

File: utils/datasets.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+            train_y = self.regression_function(train_x)
+            test_y = self.regression_function(test_x)
+            
+            # Convert to a Pandas DataFrame Object
+            train_df = {'x':train_x,'y':train_y}
+            test_df = {'x':test_x,'y':test_y}
+            train_df = pd.DataFrame(train_df)
+            test_df = pd.DataFrame(test_df)
+            
+            # Save it in a CSV File
```
Comment (augustasio) [inline utils/datasets.py]: Maybe it would be better to save not as CSV, but as .npy, or even torch tensors directly. I know numpy has np.save and np.load, PyTorch should have something like that too

File: .gitignore [hunk 9]
```diff
@@ -1,2 +1,4 @@
 data/
-Data/
\ No newline at end of file
+Data/
+test.ipynb
+.ipynb_checkpoints
```
Comment (augustasio) [inline .gitignore]: Maybe __pycache__ and datasets_debug.log files should be ignored too?

File: utils/datasets.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s [%(levelname)s] %(message)s",
+    handlers=[
+        logging.FileHandler("datasets_debug.log"),
+        logging.StreamHandler()
+    ]
+)
+
+class Regression(Dataset):
```
Comment (augustasio) [inline utils/datasets.py]: I think this class is redundant, [TensorDataset](https://pytorch.org/docs/stable/data.html#torch.utils.data.TensorDataset) does the same thing

File: utils/datasets.py (hunk 8 above)
Comment (augustasio) [reply utils/datasets.py]: Then we would also need less dependencies, as we could use torch to create, save and load

**Review by AugustasMacijauskas — CHANGES_REQUESTED** [review]
The test.ipynb that you sent me works. Other than that see comments

File: utils/datasets.py (hunk 2 above)
Comment (ms2892) [reply utils/datasets.py]: got it will fix the imports.

File: utils/datasets.py (hunk 3 above)
Comment (ms2892) [reply utils/datasets.py]: will fix it

File: utils/datasets.py (hunk 6 above)
Comment (ms2892) [reply utils/datasets.py]: Thanks for pointing out will fix this

File: utils/datasets.py (hunk 7 above)
Comment (ms2892) [reply utils/datasets.py]: well if you check there are default parameters used that are the same that were used in the paper. This epsilon was used everywhere and doesn't change.

Also the reason i did it was in case we wanna test it for more epsilons and check the performance. This way we can easily just change one value in the function and get the values.

File: utils/datasets.py (hunk 8 above)
Comment (ms2892) [reply utils/datasets.py]: Sure i will do it in numpy

File: .gitignore (hunk 9 above)
Comment (ms2892) [reply .gitignore]: yeah will add them as well

File: utils/datasets.py (hunk 10 above)
Comment (ms2892) [reply utils/datasets.py]: did not know that I will look into it and re-implement the dataset

File: utils/datasets.py (hunk 8 above)
Comment (ms2892) [reply utils/datasets.py]: had saved it in csv so that i could easily see the values

File: utils/datasets.py (hunk 7 above)
Comment (augustasio) [reply utils/datasets.py]: I think we are talking about different things, I am referring to the fact that in the paper it seems like it is the same sample everywhere, i.e. call np.random.normal once and store it in a separate variable, and then use it.

Comment (ms2892) [reply utils/datasets.py]: Yes i get it and it is saved in the utils folder. the code checks for the files if they are present or not. Also there is a set seed so if there aren't the files present by any chance, it would generate the same numbers and store it

Comment (ms2892) [reply utils/datasets.py]: so i'll re-run this code again and save it in npy files

**Review by AugustasMacijauskas — APPROVED** [review]
